### PR TITLE
[elastic/beats] Rabbitmq_1015: Add shovel metricset

### PR DESCRIPTION
**Pull request by ritalwar**
## What does this PR do?

This PR adds a new metric-set shovel for RabbitMQ module.

## Why is it important?

This status indicates if messages are being picked up and dropped off from the queue. Checking this plugin is included in the [RabbitMQ monitoring guide](https://www.rabbitmq.com/monitoring.html#readiness-probes).

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

Setup the RabbitMQ(if using docker image, prefer: rabbitmq:3-management) and enable the shovel plugin(rabbitmq_shovel and rabbitmq_shovel_management)
Create a shovel to test in Rabbitmq (Admin -> Shovel Management -> Add a new shovel)
Run metricbeat by enabling rabbitmq module and data should flow successfully in Kibana.

## Related issues

- Relates #https://github.com/elastic/integrations/issues/1015

### CODE REVIEW

**Review by belimawr — COMMENTED** [review]
I added some questions about the consistency of the docs/example with the implementation and a couple of small linting/formatting (I thought our linter should catch them o.O).

File: metricbeat/module/rabbitmq/_meta/config.reference.yml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 - module: rabbitmq
-  metricsets: ["node", "queue", "connection"]
+  metricsets: ["node", "queue", "connection","shovel"]
```
[the remark this answers is not in the document]
Comment (ritalwar) [reply metricbeat/module/rabbitmq/_meta/config.reference.yml]: Added exchange.

File: metricbeat/docs/modules/rabbitmq.asciidoc [hunk 2]
```diff
@@ -37,7 +37,7 @@ in <<configuration-metricbeat>>. Here is an example configuration:
 ----
 metricbeat.modules:
 - module: rabbitmq
-  metricsets: ["node", "queue", "connection"]
+  metricsets: ["node", "queue", "connection","shovel"]
```
[the remark this answers is not in the document]
Comment (ritalwar) [reply metricbeat/docs/modules/rabbitmq.asciidoc]: Updated.

File: metricbeat/module/rabbitmq/_meta/config.yml [hunk 3]
```diff
@@ -3,6 +3,7 @@
   #  - node
   #  - queue
   #  - connection
+  #  - shovel
```
[the remark this answers is not in the document]
Comment (ritalwar) [reply metricbeat/module/rabbitmq/_meta/config.yml:7]: Updated.

File: metricbeat/module/rabbitmq/shovel/data.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+package shovel
+
+import (
+	"encoding/json"
+
+	"fmt"
+
+	"github.com/elastic/beats/v7/metricbeat/mb"
+
```
[the remark this answers is not in the document]
Comment (ritalwar) [reply metricbeat/module/rabbitmq/shovel/data.go]: > I added some questions about the consistency of the docs/example with the implementation and a couple of small linting/formatting (I thought our linter should catch them o.O).

I have addressed all the comments, Thanks.

File: metricbeat/module/rabbitmq/shovel/shovel_test.go [hunk 5]
```diff
@@ -48,6 +48,21 @@ func TestFetchEventContents(t *testing.T) {
...
+	defer server.Close()
+
+	ms := mbtest.NewReportingMetricSetV2Error(t, getConfig(server.URL))
+	err := mbtest.WriteEventsReporterV2ErrorCond(ms, t, "", func(e mapstr.M) bool {
+		hasType, _ := e.HasKey("rabbitmq.shovel.type")
+		hasState, _ := e.HasKey("rabbitmq.shovel.state")
+		return hasType && hasState
+	})
+	if err != nil {
+		t.Fatal("write", err)
```
Comment (belimawr) [inline metricbeat/module/rabbitmq/shovel/shovel_test.go:62]: This error message could be more descriptive. What has failed when `err != nil`?

**Review by belimawr — CHANGES_REQUESTED** [review]

File: metricbeat/module/rabbitmq/shovel/shovel_test.go [hunk 6]
```diff
@@ -48,6 +48,21 @@ func TestFetchEventContents(t *testing.T) {
 	assert.Equal(t, "dynamic", event["type"])
 }
 
+func TestData(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (belimawr) [reply metricbeat/module/rabbitmq/shovel/shovel_test.go:51]: I see, thanks for clarifying!

**Review by belimawr — APPROVED** [review]

File: metricbeat/module/rabbitmq/shovel/_meta/fields.yml [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+      migration: true
+      description: >
+        Virtual host name with non-ASCII characters escaped as in C.
+    - name: node
+      type: alias
+      path: rabbitmq.node.name
+      migration: true
+      description: >
+        Node name.
+    - name: state
```
Comment (muthu-mps) [inline metricbeat/module/rabbitmq/shovel/_meta/fields.yml:23]: If this field refers to status property in the shovel document then can you rename the field name to status ? and the description should start as Current state of shovel...

Comment (muthu-mps) [inline metricbeat/module/rabbitmq/shovel/_meta/fields.yml:23]: If this field refers to status property in the shovel document then can you rename the field name to status ? and the description should start as Current state of shovel...

File: metricbeat/module/rabbitmq/shovel/shovel.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+// MetricSet for fetching RabbitMQ shovels metrics.
+type MetricSet struct {
+	*rabbitmq.MetricSet
+}
+
+// New creates new instance of MetricSet
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	ms, err := rabbitmq.NewMetricSet(base, rabbitmq.ShovelsPath)
+	if err != nil {
+		return nil, err
```
Comment (muthu-mps) [inline metricbeat/module/rabbitmq/shovel/shovel.go:43]: Add context to the return `err`

**Review by muthu-mps — APPROVED** [review]
LGTM!

**Review by ManojS-shetty — APPROVED** [review]
LGTM!

File: metricbeat/docs/modules_list.asciidoc [hunk 9]
```diff
@@ -249,10 +249,11 @@ This file is generated! See scripts/mage/docs_collector.go
 |<<metricbeat-metricset-prometheus-query,query>>   
 |<<metricbeat-metricset-prometheus-remote_write,remote_write>>   
 |<<metricbeat-module-rabbitmq,RabbitMQ>>     |image:./images/icon-yes.png[Prebuilt dashboards are available]    |  
-.4+| .4+|  |<<metricbeat-metricset-rabbitmq-connection,connection>>   
+.5+| .5+|  |<<metricbeat-metricset-rabbitmq-connection,connection>>   
 |<<metricbeat-metricset-rabbitmq-exchange,exchange>>   
 |<<metricbeat-metricset-rabbitmq-node,node>>   
 |<<metricbeat-metricset-rabbitmq-queue,queue>>   
+|<<metricbeat-metricset-rabbitmq-shovel,shovel>> beta[]
```
[the remark this answers is not in the document]
Comment (ritalwar) [reply metricbeat/docs/modules_list.asciidoc:260]: Initially yes.

**Pull request merged**